File: sources/LiquidSense Release/me/aquavit/liquidsense/event/EventManager.java
package me.aquavit.liquidsense.event;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class EventManager {

    private final HashMap<Class<? extends Event>, List<EventHook>> registry = new HashMap<>();

    public void registerListener(Listenable listener) {
        for (Method method : listener.getClass().getDeclaredMethods()) {
            if (method.isAnnotationPresent(EventTarget.class) && method.getParameterTypes().length == 1) {
                if (!method.isAccessible())
                    method.setAccessible(true);

                Class<? extends Event> eventClass = (Class<? extends Event>) method.getParameterTypes()[0];
                final EventTarget eventTarget = method.getAnnotation(EventTarget.class);

                final List<EventHook> invokableEventTargets = this.registry.getOrDefault(eventClass, new ArrayList<>());
                invokableEventTargets.add(new EventHook(listener, method, eventTarget));
                this.registry.put(eventClass, invokableEventTargets);
            }
        }
    }

    public void unregisterListener(final Listenable listenable) {
        for (final Map.Entry<Class<? extends Event>, List<EventHook>> entry : this.registry.entrySet()) {
            final List<EventHook> targets = entry.getValue();
            targets.removeIf(it -> it.getEventClass() == listenable);
            this.registry.put(entry.getKey(), targets);
        }
    }

    public final void callEvent(Event event) {
        List<EventHook> targets = this.registry.get(event.getClass());
        if (targets == null) return;
        for (EventHook invokableEventTarget : targets) {
            try {
                if (!invokableEventTarget.getEventClass().handleEvents() && !invokableEventTarget.isIgnoreCondition()) continue;
                invokableEventTarget.getMethod().invoke(invokableEventTarget.getEventClass(), event);
            }
            catch (Throwable throwable) {
                throwable.printStackTrace();
            }
        }
    }
}

File: sources/Kevin Reborn/kevin/event/EventManager.kt
/*
 * This program is free software: you can redistribute it and/or modify
 *   it under the terms of the GNU General Public License as published by
 *   the Free Software Foundation, either version 3 of the License, or
 *   (at your option) any later version.
 *
 *   This program is distributed in the hope that it will be useful,
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *   GNU General Public License for more details.
 *
 *   You should have received a copy of the GNU General Public License
 *   along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package kevin.event

import kevin.hud.element.elements.ConnectNotificationType
import kevin.hud.element.elements.Notification
import kevin.main.KevinClient
import net.minecraft.client.Minecraft
import java.lang.reflect.InvocationTargetException

@Suppress("UNCHECKED_CAST")
class EventManager {
    private val registry = HashMap<Class<out Event>, MutableList<EventHook>>()
    fun registerListener(listener: Listenable) {
        for (method in listener.javaClass.declaredMethods) {
            if (method.isAnnotationPresent(EventTarget::class.java) && method.parameterTypes.size == 1) {
                if (!method.isAccessible)
                    method.isAccessible = true

                val eventClass = method.parameterTypes[0] as Class<out Event>
                val eventTarget = method.getAnnotation(EventTarget::class.java)

                val invokableEventTargets = registry.getOrDefault(eventClass, ArrayList())
                invokableEventTargets.add(EventHook(listener, method, eventTarget))
                registry[eventClass] = invokableEventTargets
            }
        }
    }
    fun unregisterListener(listenable: Listenable) {
        for ((key, targets) in registry) {
            targets.removeIf { it.eventClass == listenable }

            registry[key] = targets
        }
    }
    fun callEvent(event: Event) {
        val targets = registry[event.javaClass] ?: return

        for (invokableEventTarget in targets) {
            try {
                if (!invokableEventTarget.eventClass.handleEvents() && !invokableEventTarget.isIgnoreCondition)
                    continue

                invokableEventTarget.call(event)
            } catch (throwable: Throwable) {
                throwable.printStackTrace()
                if (KevinClient.debug) {
                    if (throwable is InvocationTargetException) {
                        KevinClient.hud.addNotification(
                            Notification(
                                "Exception caught when calling ${event.javaClass.simpleName} in ${invokableEventTarget.eventClass.javaClass.simpleName}: ${throwable.targetException.message}",
                                "Debug",
                                ConnectNotificationType.Error
                            )
                        )
                        Minecraft.logger.warn("Exception caught when calling ${event.javaClass.simpleName} in listener ${invokableEventTarget.eventClass.javaClass.simpleName}:\n" +
                                throwable.targetException.stackTraceToString()
                        )
                    }
                }
            }
        }
    }
}